Implement Interop Type Map support in NativeAOT

jkoritzinsky opened the pull request:
Implement support for the [Interop Type Mapping feature in NativeAOT](https://github.com/dotnet/runtime/issues/110691) with the expected trimming behavior.

Contributes to #110691 
Fixes https://github.com/dotnet/runtime/issues/113362
Fixes https://github.com/dotnet/runtime/issues/115323

## Trimming behavior

The various type map attributes are explicitly trimmed out by ILC. Once we add trimmer support, we can change these attributes to be trimmed out by the attributes XML (depending on how we implement support in the trimmer).

### External type map

An entry in the external type map is kept if one of the two following cases is true:

1. The `trimTarget` type's `NecessaryTypeSymbol` is marked in the graph.


2. When the ILScanner is enabled, a `ScannedCastNode` node for `trimTarget` is in the graph.

We need this additional node as the whole-program optimization provided by the IL scanner can optimize checks like `if (obj is Foo)` or `if (obj.GetType() == typeof(Foo))` to `if(false)` if that's the only mention of the type. However, by spec, we want to still keep any `TypeMapAttribute` that has that type as a trim target.

If a check like the ones mentioned above is the only mention of a type and the `target` type has no relation to the `trimTarget` type, the `trimTarget` type will not be unnecessarily preserved, only the entry in the table (and the `target` type's `MaximallyConstructableType` node) will be preserved.

This node is not needed when not running the ILScanner, as we don't do this type of whole-program optimization without the scanner.

### Proxy type map

An entry in the proxy type map is kept if the `sourceType`'s `MaximallyConstructableType` node is kept.

## TypeMapping API

The dictionaries exposed by the `TypeMapping` API are implemented using the NativeAOT NativeFormat to embed them into the image.


## Tests

I've added a NativeAOT Test app to validate the behavior after publishing with the ILScanner enabled and updated the `Interop/TypeMap` test suite for the NativeAOT implementation.

I've also added a trimming test. I did notice, however, that the trimming test looks at the results of scanning, not codegen, so the exact list of things kept vs not kept differs between the NativeAOT test app and the trimming test (as the NativeAOT test app tests the output after codegen).

jkoritzinsky committed:
Track type cast targets

jkoritzinsky committed:
Scan for casts as part of the MethodBodyScanner instead

jkoritzinsky committed:
After feedback, first draft implementing external type map (compiler side)

jkoritzinsky committed:
Add first pass of external type map implementation for NAOT runtime

jkoritzinsky committed:
Force the target type in the external type map entry to be constructable.

jkoritzinsky committed:
First pass implementing the proxy type map

jkoritzinsky committed:
Fix compiler asserts

jkoritzinsky committed:
Don't root type map custom attributes

jkoritzinsky committed:
Provide a mechanism to specify NativeAOT-only test apps

jkoritzinsky committed:
Fix format writing so it actually works

jkoritzinsky committed:
Cache type map attribute type lookup

jkoritzinsky committed:
Add additional node in the IL Scanner when we see a cast that may be optimized away to root any external type map entry. These entries should be rooted even if the type is never instantiated and the corresponding target type is unrelated to the trimming type.

jkoritzinsky committed:
Enable the TypeMapApp test for NativeAOT and adjust as necessary to make it trim safe (and for now update the exception types thrown)

jkoritzinsky committed:
Split type map manager type so we don't need to pass null

jkoritzinsky committed:
Match attributes based on name. Don't care about the assembly

jkoritzinsky committed:
Hook up the type map manager to the trimming test driver

jkoritzinsky committed:
Add TrimmingTest and add one more test case in the NativeAOT test app

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs:
```diff
@@ -0,0 +1,218 @@
+                NativeParser entryParser;
+                while (!(entryParser = lookup.GetNext()).IsNull)
+                {
+                    string foundName = entryParser.GetString();
```
Is this materializing a string for every check in a given chain? 🤔

jkoritzinsky replied:
Found another method that doesn't do that (just wasn't exposed here) and switched to that.

Sergio0694 commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ExternalTypeMapEntryNode.cs:
```diff
@@ -0,0 +1,61 @@
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            return [
+                new DependencyListEntry(context.MaximallyConstructableType(TargetType), "Target type in external type map")
```
Is this for the value in an external type map entry? Why is it marked as maximally constructible? My understanding is that it would be minimally constructible (i.e. just get me the `Type` object with nothing else, just enough so I can look up attributes). Then if any attribute has something like `new TheType()` in an instance method, only then would the type actually be marked as constructible.

Eg. in CsWinRT we have external type map entries for all generic instantiations, but those would point to proxy types, which we never need to instantiate at all. They're only needed to lookup their attribute which then contains the actual marshalling logic we need to invoke.

Am I understanding this wrong? Or is this something we could potentially improve?

jkoritzinsky replied:
A `MaximallyConstructableType` is the "you can get custom attributes, allocate instances call a method" node. A `NecessaryType` won't let you get a System.Type instance that's usable.

Sergio0694 replied:
Mmh am I confusing some stuff that Michal perhaps mentioned he _wanted_ to add but that is not available just yet then? Basically I thought there was a way to get a `Type` instance that you can use strictly only to get attributes from and nothing else. I remember we hit that scenario in several cases around IDIC, for instance. Does that ring any bells? 🤔

jkoritzinsky committed:
Add warning validation

Sergio0694 commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ExternalTypeMapEntryNode.cs:
```diff
@@ -0,0 +1,61 @@
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context)
+        {
+            return [
+                new DependencyListEntry(context.MaximallyConstructableType(TargetType), "Target type in external type map")
```
Same question for the proxy type map: can we (if this isn't doing it already, the code is hard to understand for me as I'm not very familiar with ILC) make the target types in the proxy map to be minimally constructed if the entry is kept? And then only actually mark them as constructed for real if there's something else in that chain that would actually make them constructed (eg. a factory for the same type, in a method on an attribute over that same type).

jkoritzinsky committed:
Fix node name

jkoritzinsky committed:
Don't materialize a string for every entry as we search the hashtable

AaronRobinsonMSFT commented on src/tools/illink/test/Mono.Linker.Tests.Cases/Reflection/TypeMap.cs:
```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
```
/cc @sbomer

AaronRobinsonMSFT commented on src/libraries/System.Runtime.InteropServices/tests/TrimmingTests/TypeMap.cs:
```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
```
/cc @sbomer

jkoritzinsky commented on src/coreclr/tools/Common/Internal/Runtime/MetadataBlob.cs:
```diff
@@ -42,5 +42,9 @@ internal enum ReflectionMapBlob
         StaticsInfoHashtable                        = 34,
         GenericMethodsHashtable                     = 35,
         ExactMethodInstantiationsHashtable          = 36,
+
```
It looks like we don't use placeholders in this file unless it's in the middle of a group of 10.

MichalStrehovsky commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs:
```diff
@@ -0,0 +1,218 @@
+            return false;
+        }
+
+        private class ExternalTypeMapDictionary(NativeHashtable table, ExternalReferencesTable externalReferences) : IReadOnlyDictionary<string, Type>
```
Style: instance variables should be prefixed with underscores (not going to comment on all the places, this is not the only one).

jkoritzinsky replied:
According to our coding style, primary constructor parameters should be named like parameters, even though they can be used throughout the body of a type.

However, that also says that I shouldn't use them when it's hard to figure out where they're coming from, so I'll adjust some of these to use fields for ease of reading outside of an IDE.

jkoritzinsky replied:
I spoke to @stephentoub about the guidelines offline (and I need to put out a PR to the guidelines page with what he recommended).

MichalStrehovsky replied:
Depending on what the guidelines are, there might still be work to address here.

jkoritzinsky replied:
Guidelines are now updated in main.

MichalStrehovsky replied:
> Guidelines are now updated in main.

Could you make a pass to ensure this PR matches the guideline? I know the guideline says "sufficiently small" and that term is fuzzy, but I tend to look at these as de minimis classes, around dozen lines where I don't need to keep clicking the expand button in GitHub UI when reviewing. This PR still uses it for 50-line classes and that's a lot of clicking to find where a variable comes from.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/AssociatedTypeMapNode.cs:
```diff
@@ -0,0 +1,41 @@
+
+        public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFactory context) => Array.Empty<DependencyListEntry>();
+        public override IEnumerable<CombinedDependencyListEntry> SearchDynamicDependencies(List<DependencyNodeCore<NodeFactory>> markedNodes, int firstNode, NodeFactory context) => Array.Empty<CombinedDependencyListEntry>();
+        protected override string GetName(NodeFactory context) => "Associated type map";
```
The name should be unique in the graph - it shows up in graph diagnostic logs.

Not going to comment on all instances of this.

```suggestion
        protected override string GetName(NodeFactory context) => $"Associated type map {typeMapGroup}";
```

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/AssociatedTypeMapObjectNode.cs:
```diff
@@ -0,0 +1,87 @@
+                    typeMapHashTable = typeMapHashTables[typeMapGroup] = new VertexHashtable();
+                }
+
+                Vertex nameVertex = writer.GetUnsignedConstant(externalReferences.GetIndex(factory.MaximallyConstructableType(entryNode.Key)));
```
```suggestion
                Vertex keyVertex = writer.GetUnsignedConstant(externalReferences.GetIndex(factory.MaximallyConstructableType(entryNode.Key)));
```

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/InvalidAssociatedTypeMapEntry.cs:
```diff
@@ -0,0 +1,53 @@
+            }
+            else
+            {
+                return -1; // This node is always less than any other node
```
This should not be necessary, `other is InvalidExternalTypeMapNode otherNode` should be always true.

(Not going to comment on all occurrences.)

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/AssociatedTypeMapObjectNode.cs:
```diff
@@ -0,0 +1,87 @@
+                typeMapGroupHashTable.Append((uint)typeMapGroup.GetHashCode(), hashTableSection.Place(tuple));
+            }
+
+            foreach (InvalidExternalTypeMapNode invalidNode in factory.MetadataManager.GetInvalidExternalTypeMaps())
```
Should this be `GetInvalidAssociatedTypeMaps`?

(We might be missing test coverage.)

MichalStrehovsky commented:
I'm wondering if we can get away with fewer new node types. I like the dependency graph, but this is a lot of code in the nodes and to support the nodes.

For example, could we get away without `AssociatedTypeMapEntryNode`?

* Make `AssociatedTypeMapNode` depend on `MaximallyConstructableNode` without the indirection through `AssociatedTypeMapEntryNode`
* At the hashtable emission phase instead of iterating `AssociatedTypeMapEntryNode` that are in the graph, iterate over the `AssociatedTypeMapNode` that are in the graph and then iterate the individual entries in it. If `MaximallyConstructableNode(key).Marked` then also `MaximallyConstructableNode(value)` will be marked and you can emit it. If it's not marked, skip.

Similarly, did you consider folding `TypeMapManager` into `MetadataManager`?

MichalStrehovsky commented:
Copilot passed the Turing test here; number of comments is inversely proportional to the size of the PR in human reviewers.

![image](https://github.com/user-attachments/assets/511ed19d-0de8-4f64-a4b3-82f3319a56ac)

jkoritzinsky commented:
> I'm wondering if we can get away with fewer new node types. I like the dependency graph, but this is a lot of code in the nodes and to support the nodes.
> 
> For example, could we get away without `AssociatedTypeMapEntryNode`?
> 
> * Make `AssociatedTypeMapNode` depend on `MaximallyConstructableNode` without the indirection through `AssociatedTypeMapEntryNode`
> * At the hashtable emission phase instead of iterating `AssociatedTypeMapEntryNode` that are in the graph, iterate over the `AssociatedTypeMapNode` that are in the graph and then iterate the individual entries in it. If `MaximallyConstructableNode(key).Marked` then also `MaximallyConstructableNode(value)` will be marked and you can emit it. If it's not marked, skip.

I think this is doable and I have an idea for how to do it for the external type map as well. Should get rid of a number of the nodes.

> Similarly, did you consider folding `TypeMapManager` into `MetadataManager`?

I don't like folding it into MetadataManager, mainly because right now it exists quite independently from it (it's only passed as one of many root providers and passed to the compilation for emitting, and the choice of implementation is unrelated to the choice of MetadataManager implementation).

jkoritzinsky committed:
Remove the "type map entry" nodes and rename the "associated" type map to the "proxy" type map to match the API users call

jkoritzinsky committed:
Collapse MetadataBasedTypeMapManager and TypeMapManager

jkoritzinsky committed:
We know all the type map nodes up front in the AnalysisBasedTypeMapManager. Utilize that

jkoritzinsky committed:
Add an interface for each set of nodes to collapse the number of lists to maintain from 4 to 2

jkoritzinsky committed:
Refactor the concept of "optimized out observations we must not forget" into a separate compilation root provider

jkoritzinsky committed:
Move attribute check outside of UsageBasedMetadataManager

jkoritzinsky committed:
Put TypeMapStates into its own file as a top-level type

jkoritzinsky commented:
I removed the specific entry nodes and split the TypeMapManager into a UsageBased one and an AnalysisBased one and got all of my changes fully separated out from the MetadataManager type hierarchy.

jkoritzinsky committed:
Fix comparison to match pattern

AaronRobinsonMSFT commented on src/coreclr/tools/Common/Internal/NativeFormat/NativeFormatWriter.cs:
```diff
@@ -2044,12 +2052,20 @@ internal override void Save(NativeWriter writer)
         public override int GetHashCode()
         {
-            throw new NotImplementedException();
+            int hashCode = 13;
+            foreach (var entry in _Entries)
+            {
+                int value = (int)entry.Hashcode * 0x5498341 + 0x832424;
+                hashCode = hashCode * 31 + value;
+            }
```
Please document this algorithm, or preferrably can we use `HashCode`?

jkoritzinsky replied:
I think we can use `HashCode` here. I'll try.

jkoritzinsky committed:
Fix cast typo

jkoritzinsky committed:
Make diagnostic names for invalid type map nodes unique

jkotas commented on src/coreclr/tools/Common/Internal/NativeFormat/NativeFormatWriter.cs:
```diff
@@ -2044,12 +2052,20 @@ internal override void Save(NativeWriter writer)
         public override int GetHashCode()
         {
-            throw new NotImplementedException();
+            int hashCode = 13;
```
What purpose is this hashcode used for?

jkoritzinsky replied:
This hash code is needed because we try to insert the `TupleVertex` that contains this hash table instance into the "unified nodes" dictionary.

jkotas replied:
Does the hash code need to be deterministic for that? Can we just use `RuntimeHelpers.GetHashCode(object)`?

Sergio0694 commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ProxyTypeMapNode.cs:
```diff
@@ -0,0 +1,66 @@
+            foreach (var (key, value) in mapEntries)
+            {
+                entries.Add(new CombinedDependencyListEntry(
+                    context.MaximallyConstructableType(value),
```
Just double checking: in https://github.com/dotnet/runtime/issues/115323, we said that the proxy type map can also be used to effectively implement a fast type -> type lookup for any type of key type, including interface types (see https://github.com/dotnet/runtime/issues/115323#issuecomment-2889781663). In particular, we'll be using this in CsWinRT (with a dedicated type map group) specifically for the lookup of (potentially generic constructed types) interface types to IDIC implementation types. Will this already also correctly handle source types being interface types? Should we add unit tests for this specific scenario too, if not already there?

Also feel free to mark #115323 as also being resolved with this PR, assuming this scenario will already work 🙂

jkoritzinsky replied:
Yep this will work for interface types. Because of IDIC, we need to have a maximally constructable node for interface types:

https://github.com/dotnet/runtime/blob/c160acf0f2d773c6a30271c76e0addda5dcb508b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs#L311-L315

As a result, even if we were able to optimize that case above for types that can't ever return a valid result for IDIC, it would be an effective lookup solution for IDIC implementation types,

jkoritzinsky replied:
Reading through that thread again, there's some interesting test cases that Michal shared. I'll add some testing for those scenarios.

Sergio0694 replied:
Oh I see. Mmh question: can we restrict that to casts where the target object might actually be `IDynamicInterfaceCastable`? I.e. if the target object is a sealed type or a value type (or a constrained generic that's instantiated over a value type) that does not implement IDIC, presumably we could skip this and root fewer things, right?

jkoritzinsky replied:
Added tests for the expected interface experiences.

jkoritzinsky committed:
Just use reference equality and the default hash code for the hashtable. We don't care about unifying different hashtables together based on value equality.

jkoritzinsky committed:
Add some more test cases around interfaces and use nameof instead of raw strings

jkoritzinsky committed:
Add support in the illink analyzer test host for no-body classes and add support in the analyzer for the type map types (it's pretty small and would have been 90% of the work to gracefully not handle it).

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs:
```diff
@@ -983,5 +988,33 @@ public override bool IsReadOnly(FieldDesc field)
             }
         }
+
+        private sealed class OptimizedAwayObservationsRootProvider : ICompilationRootProvider
+        {
+            private HashSet<TypeDesc> _optimizedAwayCastTargets = new();
+            public OptimizedAwayObservationsRootProvider(ImmutableArray<DependencyNodeCore<NodeFactory>> markedNodes, DevirtualizationManager devirtualizationManager)
+            {
+                foreach (var node in markedNodes)
+                {
+                    if (node is ScannedCastTargetNode { Type: TypeDesc castTarget }
+                        && !devirtualizationManager.CanReferenceConstructedTypeOrCanonicalFormOfType(castTarget))
+                    {
+                        // Scanning observed a cast to castTarget, but castTarget can't be constructed.
+                        // We may completely remove the cast. If this is the only reference to castTarget,
+                        // this means that the codegen compilation would never observe castTarget.
+                        // Record it here so we can root the fact that the cast was observed.
+                        _optimizedAwayCastTargets.Add(castTarget);
+                    }
+                }
+            }
+
+            public void AddCompilationRoots(IRootingServiceProvider rootProvider)
+            {
+                foreach (var target in _optimizedAwayCastTargets)
+                {
+                    rootProvider.RootPossibleCastTarget(target, "Observed cast target");
+                }
+            }
+        }
```
This worries me a bit. The reason why we need this is because this seems to runs analysis for "what entries should be generated in the type map?" in optimized builds two times. We figure out what should be in the table during whole program analysis. Then we ensure a subset of the table makes it into the compilation phase (the PossibleCastTarget part). But what about the other NecessaryTypeNodes used in casting we saw during scanning that were not PossibleCastTarget? What if RyuJIT optimizes away the need of NecessaryTypeNode there?

It might be preferable to structure this so that the contents of the table is fixed during analysis and we just emit it as computed in whole program analysis (i.e. compute contents of the table, root everything at the beginning of compilation, and emit everything at the end).

jkoritzinsky replied:
I'll take a stab at structuring it to have the "analysis-based" type map have the list of entries fixed.

jkoritzinsky replied:
With the adjustments requested, you'll get reliable behavior 100% of the time (same entries in type map) for a given IL, even if RyuJIT optimizes away the only usage of a type from a "whole program optimization" in a way that we didn't account for (the previous impl checked for specific optimizations)

jkoritzinsky committed:
Don't try to remember optimizations. Just root whatever we scanned during IL Scanning. Also enhance the type map node interfaces to actually act as interfaces.

jkoritzinsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/InvalidExternalTypeMapNode.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal sealed class InvalidExternalTypeMapNode(TypeDesc typeMapGroup, MethodDesc throwingMethodStub) : DependencyNodeCore<NodeFactory>, ISortableNode, IExternalTypeMapNode
```
I've changed how this is structured based on another comment so now we're actually using the interface more directly and we have an additional implementation.

jkoritzinsky committed:
Add comment on TrimTarget

jkoritzinsky committed:
Fix failures in TypeMapApp test

Sergio0694 commented on src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.NativeAot.cs:
```diff
@@ -0,0 +1,218 @@
+                }
+            }
+
+            return new Dictionary<string, Type>(); // Return an empty dictionary if no valid type map is found.
```
Random thought: if you only call these APIs with a valid type map group, does ILC trim this path out entirely? If not, would it make sense to just return another instance of `ExternalTypeMapDictionary` with no content, or even just some other dummy type implementing the interface, to avoid constructing that `Dictionary<string, Type>` generic type instantiation? In theory it might save a tiny bit of size in apps that don't use `Dictionary<__Canon, __Canon>` at all? (not sure if it's worth it, just figured I'd mention it).

jkotas replied:
Or, should the compiler produce empty maps as needed and should this throw missing metadata exception instead?

jkoritzinsky committed:
Seal types

jkoritzinsky committed:
Put stubs on the generated asssembly

jkoritzinsky committed:
Don't instantiate Dictionary<TKey, TValue> on the "no map" path.

jkoritzinsky committed:
Persist maps for all requests and throw an exception when no map is available.

MichalStrehovsky reviewed (commented):
Looks reasonable to me otherwise.

jkoritzinsky committed:
Fix indentation and revert change in NativeFormatWriter

Sergio0694 commented on src/tools/illink/test/Mono.Linker.Tests.Cases/Reflection/TypeMap.cs:
```diff
@@ -0,0 +1,161 @@
+using Mono.Linker.Tests.Cases.Expectations.Assertions;
+using Mono.Linker.Tests.Cases.Reflection;
+
+[assembly: TypeMap<UsedTypeMap> ("TrimTargetIsTarget", typeof (TargetAndTrimTarget), typeof (TargetAndTrimTarget))]
+[assembly: TypeMap<UsedTypeMap> ("TrimTargetIsUnrelated", typeof (TargetType), typeof (TrimTarget))]
+[assembly: TypeMap<UsedTypeMap> ("TrimTargetIsUnreferenced", typeof (UnreferencedTargetType), typeof (UnreferencedTrimTarget))]
+[assembly: TypeMapAssociation<UsedTypeMap> (typeof (SourceClass), typeof (ProxyType))]
+[assembly: TypeMapAssociation<UsedTypeMap> (typeof (I), typeof (IImpl))]
+[assembly: TypeMapAssociation<UsedTypeMap> (typeof (IInterfaceWithDynamicImpl), typeof (IDynamicImpl))]
+
+[assembly: TypeMap<UnusedTypeMap> ("UnusedName", typeof (UnusedTargetType), typeof (TrimTarget))]
+[assembly: TypeMapAssociation<UsedTypeMap> (typeof (UnusedSourceClass), typeof (UnusedProxyType))]
```
Just to verify if my understanding is correct here, here's what I'd expect:

> "TypeMap TrimTarget is kept but has no type check (shouldn't keep TargetType?)"

No. The `trimTarget` type will only root the target type if used in a type check specifically.

> "TypeMapAssociation Source type is kept, but not constructed (shouldn't keep Proxy?)"

No. Proxy types should only be kept for types which are allocated (i.e. constructed classes, boxed structs, etc.).

@jkoritzinsky is this correct?

jkoritzinsky replied:
> Just to verify if my understanding is correct here, here's what I'd expect:
> 
> > "TypeMap TrimTarget is kept but has no type check (shouldn't keep TargetType?)"
> 
> No. The `trimTarget` type will only root the target type if used in a type check specifically.

We've actually expanded this due to limitations in the tracking and how it'd make code like the following unexpectedly behave. Now we will keep the entry if `TrimTarget` is kept. I'll add another test here.

```csharp

bool IsCheck<T>(object o) => o is T;

if (IsCheck<Button>(o) != (o is Button))
{
   Console.WriteLine("Should never happen.");
}

```

> > "TypeMapAssociation Source type is kept, but not constructed (shouldn't keep Proxy?)"
> 
> No. Proxy types should only be kept for types which are allocated (i.e. constructed classes, boxed structs, etc.). Basically a proxy type should be kept iff it'd be possible to get its source type by doing `obj.GetType()` on an object. Meaning that if the source type is not actually constructed, that case would be unreachable.
> 

I'll add some tests for this.

jkoritzinsky replied:
> What if instead of looking for "is there a type check", we checked if the TrimTarget IsRelevantToVariantCasting (or whatever the equivalent is in ilc)? A type is marked RelevantToVariantCasting basically whenever it's ldtoken'ed, used as a generic parameter, used as an Array element type, or reflected over. I'm not sure what the equivalent in NativeAOT is or if the concept exists in the same way, though.

That's what we're doing. The dependency on the `NecessaryTypeSymbol` provides similar guarantees.

jkoritzinsky committed:
Add more interesting test cases.

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into typemap-ilc

jkoritzinsky commented on src/tools/illink/test/Mono.Linker.Tests.Cases/Reflection/TypeMap.cs:
```diff
@@ -0,0 +1,208 @@
+	{
+		[Kept]
+		[ExpectedWarning ("IL2057", "Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'")]
+		public static void Main (string[] args)
```
I looked into this. Supporting `Unsafe.As<T>(object)` with that being the only reference to the type is difficult, However, in that case, nothing actually observes the value as a `T`. A cast does observe if the value is a `T`, so we would preserve the entry.

If you were to use `Unsafe.As<T>` and then use the result in any way that isn't immediately casting it back to `object`, then the entry would be preserved.

Sergio0694 replied:
Mmh what would be the takeaway here though? Like, does this mean it's undefined behavior or is it fine? Eg.:

```csharp
static Foo ConvertToManaged(void* ptr)
{
    return Unsafe.As<Foo>(GetThingFromComWrappers(ptr));
}
```

Is doing this fine? Basically I'm trying to figure out if we can avoid the cost of the `(T)` cast here in a safe way 🙂

jkoritzinsky replied:
The moment you call a method that's defined on type `Foo`, you observe that the object is of that type.

When you use a cast, you also observe that the object is of type `Foo`.

If you use `Unsafe.As`, then `Foo` may only be kept if someone else tries to call something on `Foo`.

Things should work as the difference shouldn't be observable, but I wouldn't recommend using unsafe unless you absolutely need it.

AaronRobinsonMSFT replied:
> According to the spec, that should guarantee that T being a trimTarget should keep its entry

Can you point to where in the spec this is stated?

> Mmh what would be the takeaway here though? Like, does this mean it's undefined behavior or is it fine? Eg.:

This is unsafe and doesn't align with any supported interop feature set. The .NET runtime provides zero guarantees when users leverage `Unsafe` APIs. The use of `Unsafe` APIs must be based on existing program correctness to be valid, it doesn't work in the opposite direction - "`Unsafe` providing some correctness".

> Basically I'm trying to figure out if we can avoid the cost of the (T) cast here in a safe way 

The cast is the way to root the type. `Unsafe.As<T>()` is not a cast and therefore is not a rooting mechanism. 

Give us a concrete measured scenario and we can work on how best to safely reduce the cost.

Sergio0694 replied:
> "Can you point to where in the spec this is stated?"

Point 1. in the PR description for the semantics of the external type map:

> "The trimTarget type's NecessaryTypeSymbol is marked in the graph."

Type arguments root their NecessaryTypeSymbol node (if not their MaximallyConstructedType node too, given you can query for reflection stuff on them from within a method, such as doing `typeof(T).GetCustomAttribute<TAttr>()`). @MichalStrehovsky also mentioned in the type map proposal thread that types used as type arguments would have to be treated like casts for the purposes of rooting external type map entries, when they're trim targets.

> "This is unsafe and doesn't align with any supported interop feature set."

My point is that due to the reason above, the spec already considers type arguments to cause trim target types to be considered kept, and therefore their associated entries to also be kept. It just seemed a bit odd and counterintuitive that `Unsafe.As<T>()` would just be a special case exception where this rule of the spec would just not apply. But I do understand that making that work correctly here would be more work, that's fair.

> "Give us a concrete measured scenario and we can work on how best to safely reduce the cost."

Sure, I can just switch to casts for now and profile later and potentially we can revisit. My rationale for asking here was just (1) making sure I at least understood how things would work there, and (2) seeing if this would follow what `ComWrappers.ComInterfaceDispatch.FindObject` would do. That API also does `Unsafe.As` internally, presumably for the same reason I had in mind (saving unnecessary cost for potentially expensive casts, given you must already know the exact type there when calling it). That seemed analogous to scenarios where one would be using the external type map, hence my question, is all 🙂

jkoritzinsky replied:
Type arguments don't necessarily root their corresponding symbol. If the body doesn't do anything with the type argument (not even `typeof(T)`) and the call is inlined, it could be optimized away entirely as far as I can tell.

AaronRobinsonMSFT replied:
> Point 1. in the PR description for the semantics of the external type map:

This is an implementation detail for native AOT implementation and shouldn't really be discussed as the design for the Type Map. Conceptually casts and specific sets fo if checks are required. Since `Unsafe.As<T>` is neither a cast nor an if check, it shouldn't root anything. How the trimmer computes that because it is an implementation detail that is likely to change.

@jkoritzinsky Please update the mentioned design to discuss how this works from a .NET/C#/IL model as opposed to implementation of native AOT and the trimmer logic.

> that types used as type arguments would have to be treated like casts

Not for Unsafe APIs. They are by definition breaking .NET rules and unsafe. We are not building a system that further encourages unsafe API usage.

> That API also does Unsafe.As internally,

Again an implementation detail that is done by the runtime team and in SPCL. What is done in SPCL is not expected to be replicated by the broader community and when discovered outside of SPCL should generally be converted to a safe pattern.

agocke replied:
Even more explicitly: SPCL is extremely special in that it’s always tied precisely with a runtime version. We can do things that literally no one else is allowed to do, because only we can guarantee that certain implementation details hold.

agocke replied:
> @jkoritzinsky Please update the mentioned design to discuss how this works from a .NET/C#/IL model as opposed to implementation of native AOT and the trimmer logic.

We may need a kind of hybrid explanation here: the rules should be defined at the IL level, but they may need to incorporate a notion of “reachability”. That is, the mere mention of a local variable type or argument type may be enough for preservation, but only if the code is considered “reachable”, which is a trimmer concept that’s almost mostly IL based, but too complicated to inline.

jkoritzinsky replied:
I'll add a design doc that explains it at the C#/IL level.

I want to keep the description of the PR in NativeAOT terms as this PR does implement the feature for NativeAOT (and similarly for the ILLink PR, use ILLink terminology in the description).

jkoritzinsky replied:
I've attempted to add a spec to this PR. Let me know what you think!

jkoritzinsky committed:
Add a more in-depth spec at the IL level describing the type map feature

AaronRobinsonMSFT commented on docs/design/features/typemap.md:
```diff
@@ -0,0 +1,223 @@
+
+Prior to .NET 10 there were at least three (3) bespoke mechanisms for this in the .NET ecosystem:
+
+* C#/WinRT - [Built-in mappings](https://github.com/microsoft/CsWinRT/blob/master/src/WinRT.Runtime/Projections.CustomTypeMappings.tt), [Generation of vtables for AOT](https://github.com/microsoft/CsWinRT/blob/b1733e95c6d35b551fc8cf6fe04e2a0c287346dd/src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs#L1597).
```
> (https://github.com/microsoft/CsWinRT/blob/master/src/WinRT.Runtime/Projections.CustomTypeMappings.tt)

This is a non-commit based link. It will not age well if the file is renamed, removed or moved.

AaronRobinsonMSFT replied:
Same for some below.

AaronRobinsonMSFT commented on docs/design/features/typemap.md:
```diff
@@ -0,0 +1,223 @@
+
+2. Types used in a managed-to-unmanaged interop operation would use `TypeMapAssociationAttribute`
+to define a conditional link between the source and proxy type. In other words, if the
+source is kept, so is the proxy type. If Trimmer observes an explicit allocation of the source
```
```suggestion
source is kept, so is the proxy type. If the Trimmer observes an explicit allocation of the source
```

AaronRobinsonMSFT commented on docs/design/features/typemap.md:
```diff
@@ -0,0 +1,223 @@
+> Additionally, it is not guaranteed that the `TypeMapAttribute`, `TypeMapAssociationAttribute`, and `TypeMapAssemblyTargetAttribute` attributes are present in the final image after a trimming tool has been run.
+
+
+5. Trimming tools will consider calls to `TypeMapping.GetOrCreateExternalTypeMapping<>` and
+`TypeMapping.GetOrCreateProxyTypeMapping<>` as intrinsics (for example, Java via `JavaTypeMapGroup`). As a result, it is not trim-compatible to call either of these methods with non-fully-instantiated generic (such as a type argument or a type that is instantiated over a type argument).
```
Is this still true? I see these in the native AOT implementation. Are they expected to be elided in the final binary?

jkoritzinsky replied:
The methods are recognized as "intrinsic" method calls in the same way that a number of methods are intrinsics for the linker (they feed into special behavior but are still emitted into the final codegen).

AaronRobinsonMSFT commented on docs/design/features/typemap.md:
```diff
@@ -0,0 +1,223 @@
+- The generic argument to the `Activator.CreateInstance<T>` method.
+- Calls to `Type.GetType` with a constant string representing the type name.
+
+Many of these instructions that can be passed a generic parameter. In that case, the trimming tool should consider type arguments of instantiations of that type as having met one of these rules and include any entries with those types as "trim target" types.
```
```suggestion
Many of these instructions can be passed a generic parameter. In that case, the trimming tool should consider type arguments of instantiations of that type as having met one of these rules and include any entries with those types as "trim target" types.
```

jkoritzinsky committed:
PR feedback

jkoritzinsky committed:
Cleanup based on primary constructor guidance

jkoritzinsky commented:
/ba-g failures are unrelated
